### PR TITLE
Run3 hcx233: Try to make CaloSimParameters from dd4Hep

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Try to make CaloSimParameter objects from dd4hep application

#### PR validation:

Tested using internal test scripts

#### if this PR is a backport please specify the original PR:

Nothing special

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by fabiocos** [comment]
please test

**Comment by cvuosalo** [comment]
+1

**Comment by bsunanda** [comment]
@cmsbuild Hold

**Comment by bsunanda** [comment]
Hold

**Comment by cvuosalo** [comment]
hold

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h [hunk 1]
```diff
@@ -2,24 +2,34 @@
 #define HcalCommonData_CaloSimParametersFromDD_h
 
 #include "DetectorDescription/Core/interface/DDsvalues.h"
+#include "DetectorDescription/Core/interface/DDCompactView.h"
+#include <DetectorDescription/DDCMS/interface/DDCompactView.h>
```
Comment (ianna) [inline Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h]: @bsunanda - please, use ```""``` for consistency

File: Geometry/HcalCommonData/plugins/CaloSimParametersESModule.cc [hunk 2]
```diff
@@ -6,6 +6,7 @@
 #include "FWCore/Framework/interface/ESTransientHandle.h"
 #include "CondFormats/GeometryObjects/interface/CaloSimulationParameters.h"
 #include "DetectorDescription/Core/interface/DDCompactView.h"
+#include <DetectorDescription/DDCMS/interface/DDCompactView.h>
```
Comment (ianna) [inline Geometry/HcalCommonData/plugins/CaloSimParametersESModule.cc]: please, use ```""``` for consistency

**Comment by cvuosalo** [comment]
unhold

File: Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h [hunk 3]
```diff
@@ -2,24 +2,34 @@
...
   bool build(const DDCompactView*, CaloSimulationParameters&);
+  bool build(const cms::DDCompactView*, CaloSimulationParameters&);
 
 private:
+  bool buildParameters(const CaloSimulationParameters&);
   std::vector<std::string> getNames(const std::string&, const DDsvalues_type&, bool);
   std::vector<int> getNumbers(const std::string&, const DDsvalues_type&, bool);
 
-  bool fromDD4Hep_;
+  template <typename T>
```
Comment (ianna) [inline Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h]: Perhaps, a private utility function can be moved out of the class API? Also, it does not have to be templated:
```
void myPrint(const std::string& value, auto& C) {
  for( auto e : C ) {
   ...
  }
}
```

File: Geometry/HcalCommonData/src/CaloSimParametersFromDD.cc [hunk 4]
```diff
@@ -1,23 +1,20 @@
 #include "Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h"
 #include "CondFormats/GeometryObjects/interface/CaloSimulationParameters.h"
-#include "DetectorDescription/Core/interface/DDCompactView.h"
 #include "DetectorDescription/Core/interface/DDFilteredView.h"
 #include "DetectorDescription/Core/interface/DDFilter.h"
 #include "DetectorDescription/Core/interface/DDValue.h"
 #include "DetectorDescription/Core/interface/DDutils.h"
-#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <DetectorDescription/DDCMS/interface/DDFilteredView.h>
```
Comment (ianna) [inline Geometry/HcalCommonData/src/CaloSimParametersFromDD.cc]: please, use ```""```

**Review by ianna — CHANGES_REQUESTED** [review]

File: Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h (hunk 1 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h]: Done

File: Geometry/HcalCommonData/plugins/CaloSimParametersESModule.cc (hunk 2 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/plugins/CaloSimParametersESModule.cc]: Done

File: Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h (hunk 3 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h]: It is moved out of class API but "auto" gives compilation warning. So I have kept the template for the time being

File: Geometry/HcalCommonData/src/CaloSimParametersFromDD.cc (hunk 4 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/src/CaloSimParametersFromDD.cc]: Done

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@ianna @cvuosalo Please approve this PR

File: Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h [hunk 5]
```diff
@@ -2,24 +2,37 @@
...
+  for (const auto& e : vec) {
+    edm::LogVerbatim("HCalGeom") << " (" << i << ") " << e;
+    ++i;
+  }
+}
+
 class CaloSimParametersFromDD {
 public:
-  CaloSimParametersFromDD(bool fromDD4Hep);
+  CaloSimParametersFromDD() {}
```
Comment (cvuosalo) [inline Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h]:
```suggestion
  CaloSimParametersFromDD() = default;
```

Comment (bsunanda) [reply Geometry/HcalCommonData/interface/CaloSimParametersFromDD.h]: Can you approve this? I shall make this change in next PR

**Comment by bsunanda** [comment]
@fabiocos @kpedro88 Could you please merge this PR so that I can proceed with the next step

**Comment by fabiocos** [comment]
@bsunanda @ianna @cvuosalo the provided test configuration fails, although the problem does not seem related to this PR, but to recent updates in some tracker material table:

```
----- Begin Fatal Exception 30-Oct-2019 14:28:17 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'p1'
   [2] Calling method for module CaloSimParametersAnalyzer/'hpa'
   [3] Using EventSetup component CaloSimParametersESModule/'caloSimulationParameters' to make data CaloSimulationParameters/'' in record HcalParametersRcd
   [4] Using EventSetup component DDCompactViewESProducer/'' to make data DDCompactView/'' in record IdealGeometryRecord
   [5] Using EventSetup component DDDetectorESProducer/'' to make data DDDetector/'' in record IdealGeometryRecord
Exception Message:
A std::exception was thrown.
Cannot find a material referenced by name:trackermaterial:T_Copper
dd4hep: while parsing /build/fabiocos/110X/28220/CMSSW_11_0_X_2019-10-29-2300/src/Geometry/HcalAlgo/data/cms-test-ddhcalHF-algorithm.xml
dd4hep: with plugin:DD4hep_CompactLoader
----- End Fatal Exception -------------------------------------------------
```

**Comment by ianna** [comment]
@bsunanda - please replace
``` 
<Include ref="Geometry/CMSCommonData/data/materials.xml"/>
```
with
```
<Include ref='Geometry/CMSCommonData/data/materials/2021/v1/materials.xml'/>
```

**Comment by ianna** [comment]
> This is related to the change in materials.xml which now depends on trackermaterial. This is not good. I am submitting a separate PR to correct for this, Please merge this PR as it is.

There is no dependency on trackermaterial xml file in the new materials xml file.

File: Geometry/HcalAlgo/data/cms-test-ddhcalHEPhase0-algorithm.xml [hunk 6]
```diff
@@ -21,6 +21,7 @@
 
   <IncludeSection>
     <Include ref="Geometry/CMSCommonData/data/materials.xml"/>
+    <Include ref="Geometry/TrackerCommonData/data/trackermaterial.xml"/>
```
Comment (ianna) [inline Geometry/HcalAlgo/data/cms-test-ddhcalHEPhase0-algorithm.xml]: it's better to replace these two files with 
```
Geometry/CMSCommonData/data/materials/2021/v1/materials.xml
```

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@ianna @cvuosalo @fabiocos Could you approve and merge this PR now?

**Comment by bsunanda** [comment]
@ianna @cvuosalo @fabiocos Could you approve and merge this PR now?

**Comment by bsunanda** [comment]
@ianna @cvuosalo @fabiocos I am waiting for this to be integrated before proceeding with next changes for HCAL and HGCAL. Can you take care of this urgently?

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by bsunanda** [comment]
@ianna I updated 2 such configurations. The rest I shall work later. I have to work on other geometry changes for dd4hep. Please approve this now.

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by ianna** [comment]
@bsunanda please, move ```myPrint``` from .h to .cc

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@ianna I have moved the "myprint"  to .cc. Can you sign it now?

**Comment by fabiocos** [comment]
The provided test configuration runs smoothly, and I understand that @bsunanda you have compared the output with the DDD build of the geometry

**Comment by bsunanda** [comment]
@ianna @fabiocos I compared the 2 outputs and they are the same. I think I put all comments of both of you. Can you approve and merge this to IB. I shall then proceed with the next step. Could you do this urgently?

**Comment by ianna** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**